### PR TITLE
Rpan app upgrade rest endpoint

### PR DESCRIPTION
**Pull request by rosepan**
Summary of this change:

1. In REST backend: added a new method: public void  upgradeApplications(String domainUID, UpgradeApplicationsModel appsToUpgrade).
2. Created two REST models: UpgradeApplicationModel.java and UpgradeApplicationsModel.java,  representing the structured data on the POST request, a list of application infos to be upgraded.
3. Created one REST resource: UpgradeApplicationsResource.java, construct the applications info from the REST models which are passed in as parameters and pass the application info to the REST backend implementation class  RestBackendImpl.upgradeApplications() method.
4. The backend implementation calls WlsConfigRetriever.readWlsAppConfigs() to get additional application info in the domain: name, targets and sourcePath, via WebLogic REST APIs.
5. WlsConfigRetriever.readWlsAppConfigs() then parses the JSON result from the WebLogic REST APIs, construct a Map<String, WlsAppConfig> for future real upgrade process (TBD).
6. Created two unit tests: 
  - RestTest.testUpgradeApplications(): make sure the REST backend, models and resources are created correctly.
  - WlsServerConfigTest.java: make sure the JSON result parsing is correct.

### CODE REVIEW

File: src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java [hunk 1]
```diff
@@ -474,4 +476,54 @@ private ResourceBundle getResourceBundle() {
...
+      recycleClient(clientHelper, client);
+    }
+    LOGGER.exiting();
+
+  }
+
+  private void upgradeApplications(ClientHolder client, String domainUID, UpgradeApplicationsModel appsToUpgrade) {
+    authorize(client, domainUID, Operation.update);
+
+    List<Domain> domains = getDomainsList(client);
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java]: This seems inefficient.  Why list all domains just to get the single one with domainUID?

File: src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java [hunk 2]
```diff
@@ -474,4 +476,54 @@ private ResourceBundle getResourceBundle() {
...
+    Domain domain = findDomain(domainUID, domains);
+
+    String namespace = getNamespace(domainUID, domains);
+
+    // Use WLS REST API to get a list of WLS application configuration
+    Map<String, WlsAppConfig> wlsAppConfigMap = getWlsAppConfigs(client, namespace, domain, appsToUpgrade);
+    System.out.println("WlsAppConfig: " + wlsAppConfigMap);
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java]: No System.out

File: src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java [hunk 3]
```diff
@@ -474,4 +476,54 @@ private ResourceBundle getResourceBundle() {
...
+
+    // TBD  do later
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java]: These changes aren't ready for master, where we only want to put completed work.

File: src/main/java/oracle/kubernetes/operator/rest/backend/RestBackend.java [hunk 4]
```diff
@@ -63,4 +65,16 @@
...
+
+
+  /**
+   * Upgrade applications
+   * @param domainUID - the unique identifier assigned to a WebLogic domain
+   * when it was registered with the WebLogic operator.  The caller is responsible
+   * for calling isDomainUID first and not calling this method if the domain has not
+   * been registered.
+   * @param appsToUpgrade - Containing info of the application to be upgraded.
+   *                        It's provided as REST API input data.
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/backend/RestBackend.java]: What does this mean?

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+   */
+  public String getApplicationName() {
+    return applicationName;
+  }
+
+  /**
+   * Get the location of the new version of application.
+   * @return
+   */
+  public String getPatchedLocation() {
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java]: Is this relative, absolute, on the persistent volume or relative to the REST client?

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    return patchedLocation;
+  }
+
+  /**
+   * Get the location the current version of application to be backed up.
+   * @return
+   */
+  public String getBackupLocation() {
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java]: Do we default this?  Are we going to provide a roll-back operation?

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+  public List<UpgradeApplicationModel> getApplications() {
+    return applications;
+  }
+
+  public void setApplications(List<UpgradeApplicationModel> applications) {
+    this.applications = applications;
+  }
+
+  //@Override
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java]: Delete if not needed

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+ *    "backupLocation": "/shared/applications/myApp2v1.ear"}
+ *   ]}
+ *
+ */
+public class UpgradeApplicationsModel extends LinkContainerModel {
+
+  private List<UpgradeApplicationModel> applications;
+
+
+  public List<UpgradeApplicationModel> getApplications() {
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java]: javadoc

File: src/main/java/oracle/kubernetes/operator/rest/resource/UpgradeApplicationsResource.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.List;
+import java.util.Map;
+
+/**
+ * UpgradeApplicationsResource is a jaxrs resource that implements the REST api for the
+ * /operator/{version}/domains/{domainUID}/upgradeApplications path.
+ * It can be used to upgrade a list of Java EE applications in a WebLogic domain.
+ */
+
+public class UpgradeApplicationsResource extends BaseResource {
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/rest/resource/UpgradeApplicationsResource.java]: What is the return from this operation?  Is there a way to see status of the rolling?

File: src/main/java/oracle/kubernetes/operator/wlsconfig/WlsServerConfig.java [hunk 10]
```diff
@@ -11,14 +18,16 @@
 /**
  * Contains configuration of a WLS server
  * <p>
- * Copyright (c) 2017, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2017-2018, Oracle and/or its affiliates. All rights reserved.
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/wlsconfig/WlsServerConfig.java]: 2017, 2018

File: src/main/java/oracle/kubernetes/operator/wlsconfig/WlsServerConfig.java [hunk 11]
```diff
@@ -79,4 +88,159 @@ public String toString() {
...
+
+  private static String getPatchedLocation(UpgradeApplicationsModel appsToPatch, String appName) {
+    List<UpgradeApplicationModel> appList = appsToPatch.getApplications();
+    String patchedLoc = null;
+    for (UpgradeApplicationModel theApp : appList) {
+      if (theApp.getApplicationName().equals(appName)) {
+        patchedLoc = theApp.getPatchedLocation();
+        if (patchedLoc != null) {
+          if (patchedLoc.isEmpty()) {
+            throw new IllegalArgumentException(MessageKeys.APPUPGRADE_MISSING_PATCHED_LOCATION + " " + appName);
```
Comment (rjeberhard) [inline src/main/java/oracle/kubernetes/operator/wlsconfig/WlsServerConfig.java]: This looks odd.  What are you doing here?

File: src/main/resources/Operator.properties [hunk 12]
```diff
@@ -119,5 +119,11 @@ WLSKO-0117=Pod for domain with domainUID {0} in namespace {1} and with server na
 WLSKO-0118=Service for domain with domainUID {0} in namespace {1} and with server name {2} deleted; validating domain
 WLSKO-0119=Service for domain with domainUID {0} in namespace {1} and with cluster name {2} deleted; validating domain
 WLSKO-0120=Ingress for domain with domainUID {0} in namespace {1} and with cluster name {2} deleted; validating domain
-
+WLSKO-0121=Trying to retrieve WLS Application configuration from admin server. {0} ms before giving up...
```
Comment (rjeberhard) [inline src/main/resources/Operator.properties]: WLS -> WebLogic
admin server -> administration server
Application -> application

File: src/main/resources/Operator.properties [hunk 13]
```diff
@@ -119,5 +119,11 @@ WLSKO-0117=Pod for domain with domainUID {0} in namespace {1} and with server na
...
+WLSKO-0122=Got exception {0} while trying to retrieve WLS Application configuration from admin server. Will attempt again in {1} milliseconds
+WLSKO-0123=Failed to read application configurations from admin server within timeout of {0} milliseconds
```
Comment (rjeberhard) [inline src/main/resources/Operator.properties]: You use ms above but milliseconds here

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java (hunk 5 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java]: Both PatchedLocation and BackupLocation are absolute paths.

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java (hunk 6 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationModel.java]: I think eventually we should and it's part of rolling upgrade. But maybe for the first stage, we can just focus on the upgrade operation. What do you think?

File: src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java (hunk 1 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java]: These are pre-existing methods that I am reusing. It seems CallBuilder only has a listDomains(ns) method.

File: src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java (hunk 2 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java]: Fixed.

File: src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java (hunk 3 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/RestBackendImpl.java]: Right. Just want to make sure the code related to REST API looks good from the reviews. Will continue on the TBDs.

File: src/main/java/oracle/kubernetes/operator/rest/backend/RestBackend.java (hunk 4 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/backend/RestBackend.java]: Change it to:
   * @param appsToUpgrade - Containing application info that is provided by the structure data
   *                      when posting an REST request
Hope this is clearer.

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java (hunk 7 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java]: Done.

File: src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java (hunk 8 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/model/UpgradeApplicationsModel.java]: Added.

File: src/main/java/oracle/kubernetes/operator/rest/resource/UpgradeApplicationsResource.java (hunk 9 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/rest/resource/UpgradeApplicationsResource.java]: Good idea. Thinking about what to return and how ....

File: src/main/java/oracle/kubernetes/operator/wlsconfig/WlsServerConfig.java (hunk 11 above)
Comment (rosepan) [reply src/main/java/oracle/kubernetes/operator/wlsconfig/WlsServerConfig.java]: this is trying to verify if customer provides incomplete info for app upgrade when posting the upgrade operation, such as, no patchedLocation, no backupLocation, we will error this out and no further upgrade processing.

File: src/main/resources/Operator.properties (hunk 13 above)
Comment (rosepan) [reply src/main/resources/Operator.properties]: Updated to milliseconds.

File: src/main/resources/Operator.properties (hunk 12 above)
Comment (rosepan) [reply src/main/resources/Operator.properties]: Fixed.

**Pull request closed without merging**